engines/scumm/smush/smush_player.cpp

```diff
@@ -1304,4 +1278,762 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
+		return -1;
+
+	memset(_smushTracks[id].blockPtr, 127, _smushTracks[id].blockSize);
+	// Track the effective number of tracks, so that if only one fails,
+	// the others can carry on as they should
+	_smushNumTracks++;
+
+	return 0;
+}
+
+void SmushPlayer::resetAudioTracks() {
+	for (int i = 0; i < _smushNumTracks; i++) {
+		_smushTracks[i].state = TRK_STATE_INACTIVE;
+		_smushTracks[i].groupId = GRP_MASTER;
+	}
+}
+
+void SmushPlayer::setGainReductionParams(int16 gainReductionLowerBound, int16 gainReductionMultiplier) {
+	if (gainReductionLowerBound)
+		_gainReductionLowerBound = gainReductionLowerBound;
+	if (gainReductionMultiplier)
+		_gainReductionFactor = gainReductionMultiplier;
+}
+
+void SmushPlayer::setGroupVolume(int groupId, int volValue) {
+	switch (groupId) {
+	case GRP_MASTER:
+		_smushTrackVols[0] = volValue; // Master
+		break;
+	case GRP_SFX:
+		_smushTrackVols[1] = volValue; // Sfx
+		break;
+	case GRP_BKGMUS:
+		_smushTrackVols[3] = volValue; // Music
+		break;
+	case GRP_SPEECH:
+		_smushTrackVols[2] = volValue; // Voice
+		break;
+	default: // Internal groups
+		for (int i = 0; i < SMUSH_MAX_TRACKS; i++) {
+			if (_smushTracks[i].groupId == groupId)
+				_smushTracks[i].volume = volValue;
+		}
+	}
+}
+
+void SmushPlayer::handleSAUDChunk(uint8 *srcBuf, uint32 size, int groupId, int vol, int pan, int16 flags, int trkId, int index, int maxFrames) {
+	int targetTrk, diff1, diff2;
+	uint32 dataSize, mod;
+	int32 maxBlockSize, saudSize;
+	uint16 flagsAccumulator[SMUSH_MAX_TRACKS];
+
+	if (!index) {
+		saudSize = READ_BE_UINT32(&srcBuf[4]);
+
+		// Find the maximum block sizes between the current ones
+		maxBlockSize = 0;
+		for (int i = 0; i < _smushNumTracks; i++) {
+			if (_smushTracks[i].blockSize > maxBlockSize) {
+				maxBlockSize = _smushTracks[i].blockSize;
+			}
+		}
+
+		diff2 = 10000000;
+		if ((flags & TRK_TYPE_MASK) != IS_BKG_MUSIC) {
+			for (int i = 0; i < _smushNumTracks; i++) {
+				if (_smushTracks[i].blockSize - saudSize >= 0) {
+					diff1 = _smushTracks[i].blockSize - saudSize;
+				} else {
+					diff1 = saudSize - _smushTracks[i].blockSize;
+				}
+
+				if (diff1 < diff2 && saudSize < 3 * _smushTracks[i].blockSize / 2) {
+					if (_smushTracks[i].blockSize - saudSize >= 0) {
+						diff2 = _smushTracks[i].blockSize - saudSize;
+					} else {
+						diff2 = saudSize - _smushTracks[i].blockSize;
+					}
+
+					maxBlockSize = _smushTracks[i].blockSize;
+				}
+			}
+		}
+
+		for (int i = 0; i < _smushNumTracks; i++) {
+			flagsAccumulator[i] = 0;
+			if (_smushTracks[i].blockSize == maxBlockSize) {
+				if (_smushTracks[i].state == TRK_STATE_INACTIVE ||
+					_smushTracks[i].state == TRK_STATE_ENDING   ||
+					_smushTracks[i].flags <= flags) {
+					if (_smushTracks[i].state == TRK_STATE_INACTIVE)
+						flagsAccumulator[i] = 0x1000;
+
+					if (_smushTracks[i].state == TRK_STATE_ENDING)
+						flagsAccumulator[i] += 0x200;
+
+					if (_smushTracks[i].flags == flags)
+						flagsAccumulator[i] += 0x400;
```

Comment:
why flags use "+" ?

Reply:
It's a mislabeling on my end, I guess I have to find a better name for that array which is an accumulator whose value depends on the flags set  :)